dev/ring-buffers.lisp

```diff
@@ -9,48 +9,119 @@
 ;;;   insert-item, dequeue, empty!, empty-p, size, total-size, first-element
 
 (defclass* ring-buffer (abstract-queue
-                          bounded-container-mixin
-                          iteratable-container-mixin
-                          concrete-container)
-  ((contents :initarg :contents
-             :reader contents)
-   (buffer-start :initform 0
-                 :reader buffer-start)
-   (buffer-end :initform 0
-               :reader buffer-end)
-   (total-size :initarg :total-size
-               :reader total-size)))
-
-(defun make-ring-buffer (size)
-  (make-instance 'ring-buffer
-    :contents (make-array size)
-    :total-size size))
+                        bounded-container-mixin
+                        iteratable-container-mixin
+                        concrete-container)
+  ((contents :reader contents
+             :initarg :contents)
+   (buffer-start :reader buffer-start
+                 :initform 0)
+   (buffer-end :reader buffer-end
+               :initform 0)
+   (total-size :reader total-size
+               :initarg :total-size)))
+
+(defclass* ring-buffer-reverse (ring-buffer) ())
+
+(defun make-ring-buffer (size &optional last-in-first-out)
+  (make-instance (if last-in-first-out 'ring-buffer-reverse 'ring-buffer)
+                 :contents (make-array size)
+                 :total-size size))
 
 (defmethod make-container ((class (eql 'ring-buffer)) &rest args)
-   (let ((total-size (getf args :total-size 1)))
-     (remf args :total-size)
-     (make-ring-buffer total-size)))
+  (let ((total-size (getf args :total-size 1))
+        (last-in-first-out (getf args :last-in-first-out nil)))
+    (remf args :total-size)
+    (remf args :last-in-first-out)
+    (make-ring-buffer total-size last-in-first-out)))
 
 
 ;;?? the (first indexes) is odd...
 (defmethod item-at ((container ring-buffer) &rest indexes)
-   (declare (dynamic-extent indexes))
-   (svref (contents container)
-          (mod (first indexes) (total-size container))))
+  (declare (dynamic-extent indexes))
+  (let ((indexes (mapcar #'(lambda (index) (fifo-index container index))
+                         indexes)))
+    (svref (contents container)
+           (mod (first indexes) (total-size container)))))
+
+(defmethod item-at ((container ring-buffer-reverse) &rest indexes)
+  (declare (dynamic-extent indexes))
+  (let ((indexes (mapcar #'(lambda (index) (lifo-index container index))
+                         indexes)))
+    (svref (contents container)
+           (mod (first indexes) (total-size container)))))
 
-(defmethod item-at! ((container ring-buffer) value &rest indexes)
+(defmethod item-at! ((container ring-buffer-reverse) value &rest indexes)
```

Comment:
I think the specialization should be against ring-buffer, not ring-buffer-reverse here (it's already there below).  Anyways, we might not need 2 item-at* functions, see below.

Reply:
What do you mean the specialization against ring-buffer? Also, yes, we don't need two item-at* functions. I can't believe I wrote multiple versions of the same method only differing by the index function. Obviously poor quality, oops.

Reply:
Both `item-at!` methods specialized on ring-buffer-reverse.  One should have specialized on `ring-buffer`.